resolver/resolver.go

```diff
@@ -467,28 +472,29 @@ func truncate(m *dns.Msg, udp bool) *dns.Msg {
 
 	furtherTruncation := m.Len() > max
 	m.Truncated = m.Truncated || furtherTruncation
-
+	if !setTruncateBit {
+		// Clear the Truncate bit regardless of whether this message used
+		// to have it set set or whether we are going to truncate it here.
+		m.Truncated = false
+	}
 	if !furtherTruncation {
 		return m
 	}
-
 	m.Extra = nil // Drop all extra records first
 	if m.Len() < max {
+		// Dropping the extra records shrunk the message down sufficiently.
 		return m
 	}
 	answers := m.Answer[:]
-	left, right := 0, len(m.Answer)
-	for {
-		if left == right {
-			break
-		}
-		mid := (left + right) / 2
-		m.Answer = answers[:mid]
-		if m.Len() < max {
-			left = mid + 1
-			continue
-		}
-		right = mid
+	search := func(n int) bool {
+		m.Answer = answers[:n]
+		return m.Len() > max
+	}
+	limit := sort.Search(len(answers), search)
+	if limit > 0 {
+		m.Answer = answers[:limit-1]
```

Comment:
Thanks for pushing back on this. Reading the docs too quickly it seems. I first (wrongly) interpreted `limit` to be an index, not "some number of messages".